_posts/2023-08-14-clear-specification-and-implementation.markdown

```diff
@@ -0,0 +1,610 @@
+---
+layout: post
+title:  "Clear Separation of Specification and Implementation in Dafny"
+date:   2023-08-14 00:00:00 +0100
+author: Aaron Tomb
+categories:
+---
+
+Dafny allows you to describe computations in several ways,
+primarily pure functions and imperative methods, each with its
+own tradeoffs. Because of these tradeoffs, it can be useful to
+combine mechanisms and relate them to each other. This post
+describes one particularly effective technique for doing so, by
+clearly separating a _specification_ from an _implementation_,
+proving properties of the specification, and proving equivalence (or
+refinement) between the specification and the implementation.
+
+To get started, let's go over the tools Dafny provides. The two most
+common computational descriptions in Dafny are terminating,
+deterministic functions and imperative, stateful methods.
+
+By writing a `function`, you can describe a mathematical object that
+maps each input deterministically to a single output. Functions are
+typically concise, and, given their genesis as one of the
```

Comment:
I don't think it is because of their genesis that they are easy to reason about. It is because of referential transparency.

Reply:
Maybe I should reword it, but what I mean to say is that mathematicians care about being able to reason about things easily, and that's why they've defined functions the way they have.